interfaces/adfsuite/amsworker.py

```diff
@@ -1197,31 +1195,7 @@ def _check_process(self):
             return False
 
     def _flatten_arrays(self, d):
-        out: Dict = {}
-        # import numpy as np
-        for key, val in d.items():
-
-            if (isinstance(val, collections.abc.Sequence) or isinstance(val, np.ndarray)) and not isinstance(val, str):
-                array = np.asarray(val)
-                if array.dtype == np.float32:
-                    # Workaround py-ubjson not knowing how to encode float32 yet
-                    array = array.astype(np.float64)
-                out[key] = array.ravel()
-                out[key + "_dim_"] = array.shape[::-1]
-                if issubclass(out[key].dtype.type, Integral) or out[key].dtype.type == np.bool_:
-                    # ubjson does not know how to convert numpy arrays of int/bool, so convert back to a list here
-                    out[key] = out[key].tolist()
-                elif out[key].dtype == np.bool_:
-                    out[key] = out[key].tolist()
-            elif isinstance(val, collections.abc.Mapping):
-                out[key] = self._flatten_arrays(val)
-            elif isinstance(val, np.float32):  # for scalar float32
-                out[key] = np.float64(val)
-            elif isinstance(val, np.int32):
-                out[key] = int(val)
-            else:
-                out[key] = val
-        return out
+        return flatten_arrays(d)
```

Comment:
Something still feels a but off here, in the separation of what lives in amspipe and plams e.g. having `flatten_arrays` in amspipe but not `unflatten_arrays`.

How is this currently working, in that how would a `AMSPipeRuntimeError` be raised currently is we are not importing/using `amspipe`? In that, is an `ASEPipeWorker` currently spun up and called somehow?

Reply:
No, the ASEPipeWorker is for the opposite direction of the pipe protocol. It is not used by any PLAMS classes.

The AMSPipeRuntimeError is raised in self._call using AMSPipeError.from_message().

One could also move unflatten_arrays into scm.amspipe.utils, there is no harm in that.

Reply:
OK thanks I see it now. I feel like this `amspipe` dependency could be then refactored to be completely removed?

Yes you would need to duplicate the `flatten_array` logic in both `plams` and `amspipe`, but then they would be totally decoupled. Then you would also just need a plams exception which had the message status that could replace the one usage of `except AMSPipeRuntimeError as exc`?

Otherwise, doesn't this change introduce an issue similar to the `InputParser`, whereby people can't pip install plams in a different python env and use the `AMSWorker`, without copying over the `amspipe` package?

Reply:
Or are you saying AMS users can in fact also easily install scm.amspipe outside of the ams python stack this is not an issue?

Reply:
AMS users can easily install scm.amspipe into any python env, they just pip install $AMSHOME/scripting/scm/amspipe.

Perhaps it's doable to have duplicate implementions of the errors and flatten_array, but it would also be nice to have a single implentation.

Reply:
Then in that case I think this is better. I am working on improving the error/help messages when a package is missing, so can easily just tell the user to do that